zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/changes/StartPartitionScaleUpApplierTest.java

```diff
@@ -0,0 +1,223 @@
+import io.camunda.zeebe.test.util.asserts.EitherAssert;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.assertj.core.api.InstanceOfAssertFactories;
+import org.junit.jupiter.api.Test;
+
+final class StartPartitionScaleUpApplierTest {
+  private final PartitionScalingChangeExecutor executor =
+      mock(PartitionScalingChangeExecutor.class);
+  private final ClusterConfiguration initialConfiguration =
+      ClusterConfiguration.init()
+          .addMember(
+              MemberId.from("1"),
+              MemberState.initializeAsActive(
+                  Map.of(
+                      1,
+                      PartitionState.active(1, DynamicPartitionConfig.init()),
+                      2,
+                      PartitionState.active(1, DynamicPartitionConfig.init()),
+                      3,
+                      PartitionState.active(1, DynamicPartitionConfig.init()))));
+
+  @Test
+  void shouldFailOnInvalidPartitionCounts() {
+    EitherAssert.assertThat(
+            new StartPartitionScaleUpApplier(executor, 0).init(initialConfiguration))
+        .left()
+        .asInstanceOf(InstanceOfAssertFactories.throwable(IllegalArgumentException.class))
+        .hasMessageContaining("Desired partition count must be greater than 1");
+
+    EitherAssert.assertThat(
+            new StartPartitionScaleUpApplier(executor, 100000).init(initialConfiguration))
+        .left()
+        .asInstanceOf(InstanceOfAssertFactories.throwable(IllegalArgumentException.class))
+        .hasMessageContaining("Desired partition count must not exceed 8192");
+  }
+
+  @Test
+  void shouldFailWhenRoutingStateIsMissing() {
+    // given - routing state not initialized
+    assertThat(initialConfiguration).returns(Optional.empty(), ClusterConfiguration::routingState);
+
+    // when - trying to init the operation
+    final var result = new StartPartitionScaleUpApplier(executor, 3).init(initialConfiguration);
+
+    // then - init fails with a useful exception
+    EitherAssert.assertThat(result)
+        .left()
+        .asInstanceOf(InstanceOfAssertFactories.throwable(IllegalStateException.class))
+        .hasMessageContaining("Routing state is not initialized yet");
+  }
+
+  @Test
+  void shouldFailWhenPartitionsAreNotAvailable() {
+    // given - cluster does not contain partition 4
+    assertThat(initialConfiguration.hasPartition(4)).isFalse();
+
+    // when - trying to init the operation to scale up to 4 partitions
+    final var result = new StartPartitionScaleUpApplier(executor, 4).init(initialConfiguration);
+
+    // then - init fails with a useful exception
+    EitherAssert.assertThat(result)
+        .left()
+        .asInstanceOf(InstanceOfAssertFactories.throwable(IllegalStateException.class))
+        .hasMessageContaining("Partition 4 is not known");
+  }
+
+  @Test
+  void shouldFailWhenRequestHandlingIsUnexpected() {
```

Comment:
:wrench: For completeness, we should also test the third case (`Mixed`). Can be a parameterized test (though unlike enums, I don't think we can ensure exhaustive use of sealed interface types here)

Reply:
Mixed is removed now :tada: